packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,278 @@
+    ) {
+        this.statusChangeEmitter = new Emitter<ConnectionStatusChangeEvent>();
+        this.endpointUrl = new Endpoint().getRestUrl().toString();
+        this.connectionState = new ConnectionStateMachine({ threshold: this.options.retry });
+    }
+
+    onStart() {
+        this.timer = window.setInterval(() => {
+            const handle = (success: boolean) => {
+                this.connectionState = this.connectionState.next(success);
+                this.statusChangeEmitter.fire(this.connectionState);
+            };
+            const xhr = new XMLHttpRequest();
+            xhr.timeout = this.options.requestTimeout;
+            xhr.onreadystatechange = event => {
+                const { readyState, status } = xhr;
+                if (readyState === XMLHttpRequest.DONE) {
+                    handle(status === 200);
+                }
+            };
+            xhr.onerror = () => handle(false);
+            xhr.ontimeout = () => handle(false);
+            xhr.open('GET', `${this.endpointUrl}/alive`);
+            try {
+                xhr.send();
+            } catch {
+                handle(false);
+            }
+        }, this.options.retryInterval);
+        this.statusChangeEmitter.fire(this.connectionState);
+    }
+
+    onStop() {
+        if (this.timer !== undefined) {
+            window.clearInterval(this.timer);
+            this.timer = undefined;
+        }
+    }
+
+    get onStatusChange() {
+        return this.statusChangeEmitter.event;
+    }
+
+}
+
+/**
+ * Just in case we need to support a bit more sophisticated state transitions than having connection or not.
+ * For instance, `pending`, `reconnecting`.
+ */
+class ConnectionStateMachine implements ConnectionStatusChangeEvent {
+
+    static readonly MAX_HISTORY = 100;
+
+    public readonly health: number;
+
+    constructor(
+        private readonly props: { readonly threshold: number },
+        public readonly state: ConnectionState = ConnectionState.ONLINE,
+        private readonly history: boolean[] = []) {
+
+        if (this.state === ConnectionState.OFFLINE) {
+            this.health = 0;
+        } else {
+            this.health = this.history.length === 0 ? 100 : Math.round((this.history.filter(success => success).length / this.history.length) * 100);
+        }
+    }
+
+    next(success: boolean): ConnectionStateMachine {
+        const newHistory = this.updateHistory(success);
+        // Initial optimism.
+        let online = true;
+        if (newHistory.length > this.props.threshold) {
+            online = newHistory.slice(-this.props.threshold).some(s => s);
+        }
+        // Ideally, we do not switch back to online if we see any `true` items but, let's say, after three consecutive `true`s.
+        return new ConnectionStateMachine(this.props, online ? ConnectionState.ONLINE : ConnectionState.OFFLINE, newHistory);
+    }
+
+    private updateHistory(success: boolean) {
```

Comment:
minor: i would prefer `protected` as default to let subclasses access or override it